### PR TITLE
provider - support for the `recover_soft_deleted_backup_protected_vm` feature

### LINKED ISSUE #24051: Support for azurerm_backup_protected_vm features backup_vault recover_soft_deleted_backup
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

Currently when a machine goes into soft delete in a backup vault terraform cannot configure it. 

already exists - to be managed via Terraform this resource needs to be imported into the State

### New or Affected Resource(s)/Data Source(s)

azurerm_backup_protected_vm

### Potential Terraform Configuration

```hcl
provider "azurerm" {
  features {
    backup_vault {
      recover_soft_deleted_backup  = true
    }
  }
  skip_provider_registration = true
}

resource "azurerm_backup_protected_vm" "protected_intent_azurerm" {
  resource_group_name = data.azurerm_recovery_services_vault.vault[0].resource_group_name
  recovery_vault_name = data.azurerm_recovery_services_vault.vault[0].name
  source_vm_id        = local.virtual_machine_id
  backup_policy_id    = local.support_backup_policy_id
}
```

### References

_No response_

### PR DESCRIPTION
**Pull request by ziyeqf**
fix #24051 

features expand
---
```
❯ go test -v ./internal/provider/ -run=TestExpand                      
=== RUN   TestExpandFeatures
    features_test.go:364: [DEBUG] Test Case: "Empty Block"
    features_test.go:364: [DEBUG] Test Case: "Complete Enabled"
    features_test.go:364: [DEBUG] Test Case: "Complete Disabled"
--- PASS: TestExpandFeatures (0.00s)
=== RUN   TestExpandFeaturesApiManagement
    features_test.go:434: [DEBUG] Test Case: "Empty Block"
    features_test.go:434: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted Api Management Enabled"
    features_test.go:434: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted Api Management Disabled"
--- PASS: TestExpandFeaturesApiManagement (0.00s)
=== RUN   TestExpandFeaturesAppConfiguration
    features_test.go:504: [DEBUG] Test Case: "Empty Block"
    features_test.go:504: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted App Configuration Enabled"
    features_test.go:504: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted App Configuration Disabled"
--- PASS: TestExpandFeaturesAppConfiguration (0.00s)
=== RUN   TestExpandFeaturesApplicationInsights
    features_test.go:569: [DEBUG] Test Case: "Empty Block"
    features_test.go:569: [DEBUG] Test Case: "Disable Generated Rule"
    features_test.go:569: [DEBUG] Test Case: "Enable Generated Rule"
--- PASS: TestExpandFeaturesApplicationInsights (0.00s)
=== RUN   TestExpandFeaturesCognitiveServices
    features_test.go:634: [DEBUG] Test Case: "Empty Block"
    features_test.go:634: [DEBUG] Test Case: "Purge on Destroy Enabled"
    features_test.go:634: [DEBUG] Test Case: "Purge on Destroy Disabled"
--- PASS: TestExpandFeaturesCognitiveServices (0.00s)
=== RUN   TestExpandFeaturesKeyVault
    features_test.go:739: [DEBUG] Test Case: "Empty Block"
    features_test.go:739: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted Key Vaults Enabled"
    features_test.go:739: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted Key Vaults Disabled"
--- PASS: TestExpandFeaturesKeyVault (0.00s)
=== RUN   TestExpandFeaturesTemplateDeployment
    features_test.go:804: [DEBUG] Test Case: "Empty Block"
    features_test.go:804: [DEBUG] Test Case: "Delete Nested Items During Deletion Enabled"
    features_test.go:804: [DEBUG] Test Case: "Delete Nested Items During Deletion Disabled"
--- PASS: TestExpandFeaturesTemplateDeployment (0.00s)
=== RUN   TestExpandFeaturesVirtualMachine
    features_test.go:922: [DEBUG] Test Case: "Empty Block"
    features_test.go:922: [DEBUG] Test Case: "Delete OS Disk Enabled"
    features_test.go:922: [DEBUG] Test Case: "Graceful Shutdown Enabled"
    features_test.go:922: [DEBUG] Test Case: "Skip Shutdown and Force Delete Enabled"
    features_test.go:922: [DEBUG] Test Case: "All Disabled"
--- PASS: TestExpandFeaturesVirtualMachine (0.00s)
=== RUN   TestExpandFeaturesVirtualMachineScaleSet
    features_test.go:1036: [DEBUG] Test Case: "Empty Block"
    features_test.go:1036: [DEBUG] Test Case: "Force Delete Enabled"
    features_test.go:1036: [DEBUG] Test Case: "Roll Instances Enabled"
    features_test.go:1036: [DEBUG] Test Case: "Scale In On Delete Disabled"
    features_test.go:1036: [DEBUG] Test Case: "All Fields Disabled"
--- PASS: TestExpandFeaturesVirtualMachineScaleSet (0.00s)
=== RUN   TestExpandFeaturesLogAnalyticsWorkspace
    features_test.go:1100: [DEBUG] Test Case: "Empty Block"
    features_test.go:1100: [DEBUG] Test Case: "Permanent Delete Enabled"
    features_test.go:1100: [DEBUG] Test Case: "Permanent Delete Disabled"
--- PASS: TestExpandFeaturesLogAnalyticsWorkspace (0.00s)
=== RUN   TestExpandFeaturesResourceGroup
    features_test.go:1165: [DEBUG] Test Case: "Empty Block"
    features_test.go:1165: [DEBUG] Test Case: "Prevent Deletion If Contains Resources Enabled"
    features_test.go:1165: [DEBUG] Test Case: "Prevent Deletion If Contains Resources Disabled"
--- PASS: TestExpandFeaturesResourceGroup (0.00s)
=== RUN   TestExpandFeaturesRecoveryServicesVault
    features_test.go:1230: [DEBUG] Test Case: "Empty Block"
    features_test.go:1230: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted Protected VM Enabled"
    features_test.go:1230: [DEBUG] Test Case: "Purge Soft Delete On Destroy and Recover Soft Deleted Protected VM Disabled"
--- PASS: TestExpandFeaturesRecoveryServicesVault (0.00s)
=== RUN   TestExpandFeaturesManagedDisk
    features_test.go:1295: [DEBUG] Test Case: "Empty Block"
    features_test.go:1295: [DEBUG] Test Case: "No Downtime Resize Enabled"
    features_test.go:1295: [DEBUG] Test Case: "No Downtime Resize Disabled"
--- PASS: TestExpandFeaturesManagedDisk (0.00s)
=== RUN   TestExpandFeaturesSubscription
    features_test.go:1343: [DEBUG] Test Case: "Empty Block"
    features_test.go:1343: [DEBUG] Test Case: "No Downtime Resize Enabled"
--- PASS: TestExpandFeaturesSubscription (0.00s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/provider	
```

### CODE REVIEW

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/features/user_flags.go [hunk 1]
```diff
@@ -79,3 +80,7 @@ type AppConfigurationFeatures struct {
 type SubscriptionFeatures struct {
 	PreventCancellationOnDestroy bool
 }
+
+type RecoveryServicesVault struct {
+	RecoverSoftDeletedBackupProtected bool
```
[the remark this answers is not in the document]
Comment (ziyeqf) [reply internal/features/user_flags.go]: renamed

File: internal/provider/features_test.go [hunk 2]
```diff
@@ -1151,6 +1170,71 @@ func TestExpandFeaturesResourceGroup(t *testing.T) {
...
+				},
+			},
+			Expected: features.UserFeatures{
+				RecoveryServicesVault: features.RecoveryServicesVault{
+					RecoverSoftDeletedBackupProtected: true,
+				},
+			},
+		},
+		{
+			Name: "Purge Soft Delete On Destroy and Recover Soft Deleted Protected VM Enabled",
```
[the remark this answers is not in the document]
Comment (ziyeqf) [reply internal/provider/features_test.go]: removed...

File: internal/services/recoveryservices/backup_protected_vm_resource.go [hunk 3]
```diff
@@ -417,6 +426,43 @@ func expandDiskLunList(input []interface{}) []interface{} {
 	return result
 }
 
+func resourceRecoveryServicesVaultBackupProtectedVMrecoverSoftDeleted(ctx context.Context, client *protecteditems.ProtectedItemsClient, opClient *backup.ProtectedItemOperationResultsClient, id protecteditems.ProtectedItemId, model *protecteditems.ProtectedItemResource) (isSoftDeleted bool, err error) {
```
[the remark this answers is not in the document]
Comment (ziyeqf) [reply internal/services/recoveryservices/backup_protected_vm_resource.go]: renamed

**Comment by ziyeqf** [comment]
TC test triggered

File: internal/services/recoveryservices/backup_protected_vm_resource.go [hunk 4]
```diff
@@ -200,6 +212,11 @@ func resourceRecoveryServicesBackupProtectedVMRead(d *pluginsdk.ResourceData, me
 	if model := resp.Model; model != nil {
 		if properties := model.Properties; properties != nil {
 			if vm, ok := properties.(protecteditems.AzureIaaSComputeVMProtectedItem); ok {
+
+				if vm.IsScheduledForDeferredDelete != nil && *vm.IsScheduledForDeferredDelete {
+					return fmt.Errorf("reading %s: the VM has been soft deleted, to recover it, specify the `recover_soft_deleted_backup_protected_vm` feature.", id)
```
Comment (magodo) [inline internal/services/recoveryservices/backup_protected_vm_resource.go:217]: Shall we do the same as if the resource is 404?

File: internal/services/recoveryservices/backup_protected_vm_resource.go [hunk 5]
```diff
@@ -105,8 +106,19 @@ func resourceRecoveryServicesBackupProtectedVMCreateUpdate(d *pluginsdk.Resource
...
+			isSoftDeleted := false
+			if meta.(*clients.Client).Features.RecoveryServicesVault.RecoverSoftDeletedBackupProtectedVM {
+				isSoftDeleted, err = resourceRecoveryServicesVaultBackupProtectedVMRecoverSoftDeleted(ctx, client, opClient, id, existing.Model)
+				if err != nil {
+					return fmt.Errorf("recovering soft deleted %s: %+v", id, err)
+				}
+			}
+
+			if !isSoftDeleted {
+				return tf.ImportAsExistsError("azurerm_backup_protected_vm", id.ID())
```
Comment (magodo) [inline internal/services/recoveryservices/backup_protected_vm_resource.go:118]: For soft deleted resource that exists, we shall error out when users haven't opt in the soft delete feature, and ensure we provide a detailed error message as: https://github.com/hashicorp/terraform-provider-azurerm/blob/2d48f47beb482c40e30b12a1f77e95002e0b6cbe/internal/services/keyvault/key_vault_resource.go#L276-L284

**Review by magodo — CHANGES_REQUESTED** [review]

**Review by magodo — COMMENTED** [review]
LGTM now!

**Comment by ziyeqf** [comment]
just a comment to remove the `waiting-response` label

File: internal/provider/features.go [hunk 6]
```diff
@@ -255,6 +255,21 @@ func schemaFeatures(supportLegacyTestSuite bool) *pluginsdk.Schema {
 			},
 		},
 
+		"recovery_services_vault": {
```
Comment (katbyte) [inline internal/provider/features.go:258]: this should be plural 
```suggestion
		"recovery_services_vaults": {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

**Review by katbyte — CHANGES_REQUESTED** [review]
We have a bunch of test failures:

can we please fix up all tests?

image

**Comment by ziyeqf** [comment]
The service seems to have something wrong and needs workaround on https://github.com/hashicorp/terraform-provider-azurerm/pull/24978

**Review by katbyte — CHANGES_REQUESTED** [review]
https://github.com/hashicorp/terraform-provider-azurerm/pull/24978 has been merged but i am still seeing test failures 
image

**Comment by ziyeqf** [comment]
> #24978 has been merged but i am still seeing test failures image

Code updated and the tests pass now. I removed the change on `site_recovery_xxx` files, as it should not be involved.
image

File: website/docs/guides/features-block.html.markdown [hunk 7]
```diff
@@ -199,6 +205,12 @@ The `resource_group` block supports the following:
 
 ---
 
+The `recovery_services_vault` block supports the following:
+
+* `recover_soft_deleted_backup_protected_vm` - (Optional) Should the `azurerm_backup_protected_vm` resource recover a Soft-Deleted protected VM? Defaults to `true`.
```
Comment (katbyte) [inline website/docs/guides/features-block.html.markdown:210]: I think this should default to false?

File: internal/provider/features.go [hunk 8]
```diff
@@ -260,6 +260,21 @@ func schemaFeatures(supportLegacyTestSuite bool) *pluginsdk.Schema {
...
+		"recovery_services_vaults": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*schema.Schema{
+					"recover_soft_deleted_backup_protected_vm": {
+						Type:     pluginsdk.TypeBool,
+						Optional: true,
+						Default:  true,
```
Comment (katbyte) [inline internal/provider/features.go:272]: this is going to change the existing behaviour?
```suggestion
						Default:  false,
```

**Review by katbyte — CHANGES_REQUESTED** [review]
the tests for this also still seem quite flakey?
image

**Comment by ziyeqf** [comment]
> recover_soft_deleted_backup_protected_vm

TBH these tests on main branch are also flakey..

**Review by katbyte — APPROVED** [review]
This doesn't seem to be a flakey test failure:

```
------- Stdout: -------
=== RUN   TestAccBackupProtectedVm_updateBackupPolicyId
=== PAUSE TestAccBackupProtectedVm_updateBackupPolicyId
=== CONT  TestAccBackupProtectedVm_updateBackupPolicyId
    testcase.go:113: Step 1/5 error: Error running pre-apply refresh: exit status 1
        
        Error: Duplicate provider configuration
        
          on terraform_plugin_test.tf line 26:
          26: provider "azurerm" {
        
        A default (non-aliased) provider configuration for "azurerm" was already
        given at terraform_plugin_test.tf:21,1-19. If multiple configurations are
        required, set the "alias" argument for alternative configurations.
    testing_new.go:79: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Duplicate provider configuration
        
          on terraform_plugin_test.tf line 26:
          26: provider "azurerm" {
        
        A default (non-aliased) provider configuration for "azurerm" was already
        given at terraform_plugin_test.tf:21,1-19. If multiple configurations are
        required, set the "alias" argument for alternative configurations.
--- FAIL: TestAccBackupProtectedVm_updateBackupPolicyId (5.44s)
FAIL
```

Should be good to merge once the test is fixed

**Comment by ziyeqf** [comment]
uh...my fault. updated
```
❯❯ tftest recoveryservices TestAccBackupProtectedVm_updateBackupPolicyId
=== RUN   TestAccBackupProtectedVm_updateBackupPolicyId
=== PAUSE TestAccBackupProtectedVm_updateBackupPolicyId
=== CONT  TestAccBackupProtectedVm_updateBackupPolicyId
--- PASS: TestAccBackupProtectedVm_updateBackupPolicyId (877.88s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/recoveryservices      879.994s
```

**Review by katbyte — APPROVED** [review]
LGTM 📨

**Pull request merged**